SeaPkg/Library/BasePeCoffValidationLib/BasePeCoffValidationLib.c

```diff
@@ -0,0 +1,451 @@
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to read memory attribute of 0x%p: 0x%x for entry at 0x%p - %r\n", __func__, AddrInTarget, MemAttrHdr->TargetMemorySize, MemAttrHdr, Status));
+    goto Done;
+  }
+
+  // Check if the memory attributes of the target image meet the requirements
+  if (((MemAttr & MemAttrHdr->TargetMemoryAttributeMustHave) != MemAttrHdr->TargetMemoryAttributeMustHave) &&
+      ((MemAttr & MemAttrHdr->TargetMemoryAttributeMustNotHave) != 0))
+  {
+    DEBUG ((DEBUG_ERROR, "%a: Current entry range 0x%p: 0x%x attribute 0x%x violated aux file specification must have 0x%x and must not have 0x%x\n", __func__, Hdr, Hdr->Size, MemAttr, MemAttrHdr->TargetMemoryAttributeMustHave, MemAttrHdr->TargetMemoryAttributeMustNotHave));
+    Status = EFI_SECURITY_VIOLATION;
+    goto Done;
+  }
+
+  Status = EFI_SUCCESS;
+
+Done:
+  return Status;
+}
+
+/**
+  Validates a specific region in the target image buffer denoted by [Hdr->Offset: Hdr->Offset + Hdr->Size]
+  matches the content in the original image buffer as specified by TargetOffset in the validation entry.
+
+  @param[in] TargetImage               The pointer to the target image buffer.
+  @param[in] Hdr                       The header of the validation entry.
+  @param[in] OriginalImageBaseAddress  The pointer to the original image buffer.
+
+  @retval EFI_SUCCESS             The target image passes the validation.
+  @retval EFI_INVALID_PARAMETER   One of the input parameters is a null pointer.
+  @retval EFI_INVALID_PARAMETER   The validation entry has an invalid signature.
+  @retval EFI_INVALID_PARAMETER   The validation entry has an invalid size.
+  @retval EFI_SECURITY_VIOLATION  The target image does not match the content in the original image buffer.
+**/
+EFI_STATUS
+EFIAPI
+PeCoffImageValidationSelfRef (
+  IN CONST VOID                           *TargetImage,
+  IN CONST IMAGE_VALIDATION_ENTRY_HEADER  *Hdr,
+  IN CONST VOID                           *OriginalImageBaseAddress
+  )
+{
+  IMAGE_VALIDATION_SELF_REF  *SelfRefHdr;
+  EFI_STATUS                 Status;
+  EFI_PHYSICAL_ADDRESS       AddrInTarget;
+  EFI_PHYSICAL_ADDRESS       AddrInOrigin;
+
+  if ((TargetImage == NULL) || (Hdr == NULL) || (OriginalImageBaseAddress == NULL)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: At least one invalid input parameter: TargetImage 0x%p, Hdr 0x%p, OriginalImageBaseAddress 0x%p\n",
+      __func__,
+      TargetImage,
+      Hdr,
+      OriginalImageBaseAddress
+      ));
+    Status = EFI_INVALID_PARAMETER;
+    goto Done;
+  }
+
+  if ((Hdr->EntrySignature != IMAGE_VALIDATION_ENTRY_SIGNATURE) || (Hdr->ValidationType != IMAGE_VALIDATION_ENTRY_TYPE_SELF_REF)) {
+    DEBUG ((DEBUG_ERROR, "%a: Invalid entry signature 0x%x or type 0x%x at 0x%p\n", __func__, Hdr->EntrySignature, Hdr->ValidationType, Hdr));
+    Status = EFI_COMPROMISED_DATA;

```

Comment:
For invalid type and signature the return status is EFI_COMPROMISED_DATA here but in every other function it's EFI_INVALID_PARAMETER.  Personally I think it should be EFI_COMPROMISED_DATA in the other functions as well but at minimum it should be consistent across them all.